packages/mui-system/src/useThemeProps/useThemeProps.js

```diff
@@ -7,6 +8,5 @@ export default function useThemeProps({ props, name, defaultTheme, themeId }) {
   if (themeId) {
     theme = theme[themeId] || theme;
   }
-  const mergedProps = getThemeProps({ theme, name, props });
-  return mergedProps;
+  return React.useMemo(() => getThemeProps({ theme, name, props }), [theme, name, props]);
```

Comment:
I wonder what's the stability of `props` itself here ?
Have you done any benchmark for this hook ?

Reply:
In the grid `props` is stable because we `React.memo` (or more exactly [`fastMemo`](https://github.com/mui/mui-x/blob/master/packages/x-data-grid/src/utils/fastMemo.ts)) most of our components including the root one, for non-memoized components it would be stable only if the component is re-rendering because one of its hooks changed, but it wouldn't be stable if it re-renders because its parent re-rendered (the most common case I think).

Now that I think, I guess it would be possible to add the `React.useMemo` outside `useThemeProps` (in the mui-x codebase), it would avoid the cost for the rest of the components. It does feel a bit cleaner to have a stable output, but it's more performant to add it only where needed. I could be convinced either way. What do you think?

Reply:
I'd also assume that most of the times, parent will re-render making the memo useless in this case.
Only of it was using some specific keys from props instead of the whole object.

Reply:
I've thought about it some more and I prefer to add the memoization only where it's needed, so I've removed this part. The only thing left is the refactor for `resolveProps`, but it's a good refactor as there were some issues with the code there, so I'll merge this PR anyway.

Reply:
I'm going to proceed with the plan above, but might need to come back. Can't wrap `useThemeProps` with `useMemo` because it is itself a hook (calling `useContext`), and needs to be reactively bound to the theme context change.

So `useMemo(() => useThemeProps({ name, props }), [name, props]))` is incorrect. I have more thinking to do because for the design-system agnostic objective for the grid I need to get rid of the dependency on `useThemeProps` anyway, so I'll do the thinking and merge this small refactor as-is.